### PR TITLE
Renter Paths

### PR DESCRIPTION
**Pull request by lukechampine**
As discussed in #894:
- Leading slashes have been removed
- The renter creates the directory structure in its persistence folder.

One possible modification here would be to add a `siaroot` folder to the renter folder, under which the .sia directory structure would be stored. This would prevent confusion stemming from the .sia files living alongside renter.json, renter.log, and hostdb.log.

### CODE REVIEW

**Comment by Mingling94** [comment]
* https://github.com/NebulousLabs/Sia/pull/891/files#r48834586
* https://github.com/NebulousLabs/Sia/pull/891/files#r48919196
* https://github.com/NebulousLabs/Sia/pull/891/files#r48987747
Does this need any other reflection in the API.md or is it fine the way it is?
I know you would do `/renter/delete/{path}` instead of `/renter/delete{/path}` so I guess that's the case?

**Comment by Mingling94** [comment]
Also to be even more RESTful couldn't these
/renter/delete/{path} [POST]
/renter/download/{path} [GET]
/renter/rename/{path} [POST]
/renter/share/{path} [GET]
/renter/shareascii/{path} [GET]
/renter/upload/{path} [POST]
/renter/hosts/active [GET]
be these
/renter/file/{path} [GET]
/renter/file/{path} [DELETE]
/renter/file/{path} [POST]

**Comment by DavidVorick** [comment]
We made a decision to only use GET and POST in our api, because for whatever reason some libraries only have good support for GET and POST request. We sacrifice some RESTfulness, but we increase compatibility and reduce the barrier to entry for interacting with our API, which I think is more important.

**Comment by lukechampine** [comment]
Important note: folders behavior may be a little strange. Namely, if your renter folder looks like `foo/bar/baz.sia`, you can still upload files named `foo` and `foo/bar`, since the `.sia` prefix will be appended before saving. In other words, the renter skips folders when checking for naming conflicts.

**Comment by DavidVorick** [comment]
What happens if I have a folder named `foo.sia`? Does the naming conflict get discovered?

**Comment by DavidVorick** [comment]
We should have tests for naming conflicts

File: api/server_helpers_test.go [hunk 1]
```diff
@@ -231,6 +232,26 @@ func (st *serverTester) coinAddress() string {
...
+		return err
+	}
+	// mine block
+	st.miner.AddBlock()
+	var hosts ActiveHosts
+	time.Sleep(1 * time.Second)
+	st.getAPI("/renter/hosts/active", &hosts)
+	if len(hosts.Hosts) == 0 {
+		return errors.New("host announcement not seen")
+	}
```
Comment (DavidVorick) [inline api/server_helpers_test.go]: can we make this a little more robust?

```
st.getAPI()
for i := 0; i < 50 && len(hosts.Hosts) == 0; i++ {
     time.Sleep(50*time.Millisecond)
    st.getAPI()
}
if len == 0 {
    return errors.New()
}

This allows us to bring the average case to a lower time, especially for faster machines, and increases tolerance for slow machines or rare situations that might take a longer time.

File: modules/renter/files.go [hunk 2]
```diff
@@ -63,7 +63,8 @@ func (f *file) chunkSize() uint64 {
 // numChunks returns the number of chunks that f was split into.
 func (f *file) numChunks() uint64 {
 	n := f.size / f.chunkSize()
-	if f.size%f.chunkSize() != 0 {
+	// last chunk may have padding
+	if f.size%f.chunkSize() != 0 || f.size == 0 {
```
Comment (DavidVorick) [inline modules/renter/files.go]: might be worth adding a comment about how this handles 0-byte files.

File: modules/renter/files_test.go [hunk 3]
```diff
@@ -6,6 +6,33 @@ import (
...
+
+		{101, 10, 1, 11}, // padded
+		{101, 10, 2, 6},  // padded
+
+		{10, 100, 1, 1}, // larger piece than file
+		{0, 10, 1, 1},   // 0-length
+	}
+
+	for _, test := range tests {
+		rsc, _ := NewRSCode(test.piecesPerChunk, 1) // can't use 0
```
Comment (DavidVorick) [inline modules/renter/files_test.go]: Can we upgrade the tests to also mess with the parity?

I'm not completely understanding this test, but I thought that the redundancy was important to the number of chunks?

File: modules/renter/persist.go [hunk 4]
```diff
@@ -195,32 +202,37 @@ func (r *Renter) save() error {
...
-	for _, path := range filenames {
-		// Skip non-sia files.
-		if filepath.Ext(path) != ShareExtension {
-			continue
-		}
-		file, err := os.Open(filepath.Join(r.persistDir, path))
+	// Recursively load all files found in renter directory. Errors
+	// encountered during loading are logged, but are not considered fatal.
+	err := filepath.Walk(r.persistDir, func(path string, info os.FileInfo, err error) error {
+		// These errors should be considered fatal.
```
Comment (DavidVorick) [inline modules/renter/persist.go]: why should these errors be considered fatal? This comment is incomplete

File: modules/renter/persist.go [hunk 5]
```diff
@@ -195,32 +202,37 @@ func (r *Renter) save() error {
...
 		if err != nil {
-			// maybe just skip?
 			return err
 		}
+
+		// Skip folders and non-sia files.
+		if info.IsDir() || filepath.Ext(path) != ShareExtension {
```
Comment (DavidVorick) [inline modules/renter/persist.go]: Do we want to eventually have support in the renter for empty folders? Seems like a good idea to me.

File: api/renter_test.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+		if filepath.Ext(path) == ".sia" {
+			rel, _ := filepath.Rel(stDir, path)
+			walkStr += rel
+		}
+		return nil
+	})
+	expWalkStr := "renter/foo/bar/test.sia"
+	if walkStr != expWalkStr {
+		t.Fatalf("Bad walk string: expected %v, got %v", expWalkStr, walkStr)
+	}
```
Comment (DavidVorick) [inline api/renter_test.go]: I think that this is the wrong approach for checking that the test succeeded. Ideally, you'd be able to verfiy without needing to check the internal state, and just use the api.

Would it work to instead query the list of files known to the renter and verify that the file returned has '/` in it?

File: modules/renter/persist_test.go [hunk 7]
```diff
@@ -161,3 +161,68 @@ func TestRenterSaveLoad(t *testing.T) {
...
+		}
+		rel, _ := filepath.Rel(rt.renter.persistDir, path) // strip testdir prefix
+		walkStr += rel
+		return nil
+	})
+	// walk will descend into foo/bar/, reading baz, bar, and finally foo
+	expWalkStr := (f3.name + ".sia") + (f2.name + ".sia") + (f1.name + ".sia")
+	if walkStr != expWalkStr {
+		t.Fatalf("Bad walk string: expected %v, got %v", expWalkStr, walkStr)
+	}
```
Comment (DavidVorick) [inline modules/renter/persist_test.go]: this check makes sense, since we're testing the renter as opposed to the api

**Comment by DavidVorick** [comment]
I would like to see a few more tests added, particularly for corner cases:

+ uploading a file when there is a name conflict
+ uploading a file when there is a name conflict with a folder
+ uploading a 0 byte file
+ uploading a file that's interpreted as having a leading slash

I think there was one more edge case that should be tested, but I've forgotten

File: modules/renter/files_test.go (hunk 3 above)
Comment (lukechampine) [inline modules/renter/files_test.go]: redundancy is relevant only with regard to how it affects the number of pieces per chunk, i.e. the number of data pieces. So the parity is being manipulated (just between 1 and 2) via `piecesPerChunk`.

File: modules/renter/persist.go (hunk 5 above)
Comment (lukechampine) [inline modules/renter/persist.go]: empty folders are even trickier than empty files. At the moment we don't support creating folders at all (only creating files within folders), so I think it's fine for now.
Later I suppose we'll add support for folder operations, and decide what makes sense then. As usual, I plan to find out how Dropbox does it.

File: api/renter_test.go (hunk 6 above)
Comment (lukechampine) [inline api/renter_test.go]: yeah, I should probably verify that `/renter/files` list the file properly and that `/renter/download/<path>` works.

File: modules/renter/persist.go (hunk 5 above)
Comment (DavidVorick) [inline modules/renter/persist.go]: ok, that's fine with me. I would like to support empty folders in the future, (as well as full file permissions, which might be how we achieve empty folders), but for now it's okay

File: api/renter_test.go (hunk 6 above)
Comment (DavidVorick) [inline api/renter_test.go]: :+1:

**Comment by lukechampine** [comment]
The behavior for the last test (uploading a file with a leading slash, e.g. `/renter/upload//foo/bar` is the same as without the slash, except that the nickname contains the leading slash. I was worried that it might create the directory structure at the root level (in `/foo`), but this doesn't happen because the renter calls `filepath.Join(r.persistDir, f.name)`, which removes the double slash in `siaDir/renter//foo/bar`.
Is this the proper behavior? I think it mostly makes sense, but it could cause some confusing behavior (like /renter/upload//foo/bar` overwriting the sia file of `/renter/upload/foo/bar`). Then again, people shouldn't be using leading slashes in the first place...

**Comment by lukechampine** [comment]
Ok, trying to create a file with a folder naming conflict results in an error: `rename foo/bar.sia_temp foo/bar.sia: is a directory`. I think this is satisfactory, especially since folders ending in ".sia" are sure to be very rare. We can address this more robustly when we add first-class folder support.

**Comment by DavidVorick** [comment]
Should return an error for files with a leading slash

File: api/renter_test.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+	if err == renter.ErrDuplicateNickname {
+		t.Fatalf("expected %v, got %v", renter.ErrDuplicateNickname, err)
+	}
+
+	// Upload using nickname that conflicts with folder.
+	err = st.stdPostAPI("/renter/upload/foo/bar", uploadValues)
+	if err == nil {
+		t.Fatal("expecting conflict error, got nil")
+	}
+}
```
Comment (DavidVorick) [inline api/renter_test.go]: :+1:
